File: typescript/src/interceptors/tool-spans.ts
/**
 * Tool-span extraction (Phase 1C).
 *
 * Provider interceptors historically emitted only a single `llm_call` span and dropped the
 * model's tool calls entirely (the most common agent-failure class — tool errors + tool
 * loops — was therefore invisible). These helpers derive structured `tool_call` and
 * `tool_result` spans from a provider request/response so tool usage is captured with NO
 * manual instrumentation.
 *
 * - `tool_call` spans come from the model's response (the calls it REQUESTED), with arguments
 *   parsed into structured JSON (not stringified into the output text).
 * - `tool_result` spans come from the tool messages the caller feeds back on the NEXT request
 *   (the verbatim recorded result, including errors/empty). They are deduped by the provider
 *   tool-call id so they are emitted once, not on every subsequent turn.
 *
 * Detectors downstream (2C schema validation, 2D loop detection, 3C tool-output hallucination)
 * depend on these spans + the `tool_call_id` linkage carried in metadata.
 */
import { SpanData, SpanType } from "../trace.js";
import { genId, nowIso, truncateJson } from "../utils.js";

type Emit = (span: SpanData) => void;

// Bounded dedup of emitted tool_result spans (keyed by provider tool-call id). Cleared when a
// new trace installs its callback (see reset call in each interceptor) to bound memory.
const emittedToolResultIds = new Set<string>();

export function resetToolResultDedup(): void {
  emittedToolResultIds.clear();
}

function markEmitted(id: string): boolean {
  if (emittedToolResultIds.has(id)) return false;
  if (emittedToolResultIds.size > 5000) emittedToolResultIds.clear();
  emittedToolResultIds.add(id);
  return true;
}

/** Parse a JSON-string arguments payload into structured JSON; leave non-strings as-is. */
export function parseToolArgs(args: unknown): unknown {
  if (typeof args !== "string") return args;
  try {
    return JSON.parse(args);
  } catch {
    return args;
  }
}

/**
 * Extract declared tool parameter schemas (name → JSON schema) from a provider request's tool
 * definitions, so the detection engine can validate tool_call arguments against ground truth.
 */
export function extractToolSchemas(provider: "openai" | "anthropic" | "gemini", tools: unknown): Record<string, unknown> | undefined {
  const out: Record<string, unknown> = {};
  if (provider === "gemini") {
    // config.tools = [{ functionDeclarations: [{ name, parameters }] }]
    if (!Array.isArray(tools)) return undefined;
    for (const group of tools as Array<{ functionDeclarations?: Array<{ name?: string; parameters?: unknown }> }>) {
      for (const fd of group?.functionDeclarations || []) {
        if (fd?.name && fd.parameters) out[fd.name] = fd.parameters;
      }
    }
  } else if (Array.isArray(tools)) {
    for (const t of tools as Array<{ name?: string; input_schema?: unknown; function?: { name?: string; parameters?: unknown } }>) {
      if (provider === "openai" && t.function?.name && t.function.parameters) out[t.function.name] = t.function.parameters;
      if (provider === "anthropic" && t.name && t.input_schema) out[t.name] = t.input_schema;
    }
  }
  return Object.keys(out).length ? out : undefined;
}

/**
 * Capture the sampling/determinism envelope from the request options so replay-divergence (2A)
 * and regression replay (2E) can compare sampling config, not just the model. Normalizes provider
 * field names to { temperature, top_p, top_k, seed, max_tokens }. Returns undefined if none set.
 */
export function extractSamplingParams(provider: "openai" | "anthropic" | "gemini", opts: unknown): Record<string, unknown> | undefined {
  const o = (opts || {}) as Record<string, unknown>;
  const cfg = (provider === "gemini" ? (o.config as Record<string, unknown> | undefined) : o) || {};
  const out: Record<string, unknown> = {};
  const put = (key: string, v: unknown) => { if (v !== undefined && v !== null) out[key] = v; };
  if (provider === "gemini") {
    put("temperature", cfg.temperature);
    put("top_p", cfg.topP);
    put("top_k", cfg.topK);
    put("seed", cfg.seed);
    put("max_tokens", cfg.maxOutputTokens);
  } else {
    put("temperature", cfg.temperature);
    put("top_p", cfg.top_p);
    put("top_k", cfg.top_k); // anthropic only
    put("seed", cfg.seed); // openai only
    put("max_tokens", cfg.max_tokens ?? cfg.max_completion_tokens);
  }
  return Object.keys(out).length ? out : undefined;
}

function toolCallSpan(name: string, input: unknown, parentId: string | null, model: string | undefined, toolCallId?: string): SpanData {
  const now = nowIso();
  return {
    id: genId(), trace_id: "", parent_id: parentId,
    span_type: SpanType.TOOL_CALL, name: name || "tool",
    ...(model ? { model } : {}),
    input: truncateJson(input),
    started_at: now, ended_at: now, duration_ms: 0,
    ...(toolCallId ? { metadata: { tool_call_id: toolCallId } } : {}),
  };
}

function toolResultSpan(name: string, output: unknown, isError: boolean, toolCallId?: string): SpanData {
  const now = nowIso();
  return {
    id: genId(), trace_id: "", parent_id: null,
    span_type: SpanType.TOOL_RESULT, name: name || "tool_result",
    output: truncateJson(output),
    started_at: now, ended_at: now, duration_ms: 0,
    ...(isError ? { error: typeof output === "string" ? output : JSON.stringify(output) } : {}),
    ...(toolCallId ? { metadata: { tool_call_id: toolCallId } } : {}),
  };
}

// ─── OpenAI (chat.completions) ───────────────────────────────────────────────
interface OAToolCall { id?: string; type?: string; function?: { name?: string; arguments?: unknown } }
interface OAMessage { role?: string; content?: unknown; tool_calls?: OAToolCall[]; tool_call_id?: string; name?: string }

/** Emit tool_call spans from an OpenAI assistant response message. */
export function emitOpenAIToolCalls(message: OAMessage | undefined, parentId: string, model: string | undefined, emit: Emit): void {
  const calls = message?.tool_calls;
  if (!Array.isArray(calls)) return;
  for (const c of calls) {
    const name = c.function?.name || "tool";
    emit(toolCallSpan(name, parseToolArgs(c.function?.arguments), parentId, model, c.id));
  }
}

/** Emit tool_result spans from OpenAI request messages (role:"tool"), deduped by tool_call_id. */
export function emitOpenAIToolResults(messages: OAMessage[] | undefined, emit: Emit): void {
  if (!Array.isArray(messages)) return;
  // Map tool_call_id -> tool name from any assistant tool_calls in the same message list.
  const nameById = new Map<string, string>();
  for (const m of messages) {
    if (m?.role === "assistant" && Array.isArray(m.tool_calls)) {
      for (const c of m.tool_calls) if (c.id) nameById.set(c.id, c.function?.name || "tool");
    }
  }
  for (const m of messages) {
    if (m?.role !== "tool" || !m.tool_call_id) continue;
    if (!markEmitted(`oa:${m.tool_call_id}`)) continue;
    const content = m.content;
    const isError = typeof content === "string" && /error|exception|failed/i.test(content);
    emit(toolResultSpan(nameById.get(m.tool_call_id) || m.name || "tool_result", content, isError, m.tool_call_id));
  }
}

// ─── Anthropic (messages) ────────────────────────────────────────────────────
interface AnthropicBlock { type?: string; name?: string; input?: unknown; id?: string; tool_use_id?: string; content?: unknown; is_error?: boolean }
interface AnthropicMessage { role?: string; content?: unknown }

/** Emit tool_call spans from Anthropic response content blocks (type:"tool_use"). */
export function emitAnthropicToolCalls(content: unknown, parentId: string, model: string | undefined, emit: Emit): void {
  if (!Array.isArray(content)) return;
  for (const block of content as AnthropicBlock[]) {
    if (block?.type !== "tool_use") continue;
    emit(toolCallSpan(block.name || "tool", block.input, parentId, model, block.id));
  }
}

/** Emit tool_result spans from Anthropic request messages (content blocks type:"tool_result"). */
export function emitAnthropicToolResults(messages: AnthropicMessage[] | undefined, emit: Emit): void {
  if (!Array.isArray(messages)) return;
  for (const m of messages) {
    if (!Array.isArray(m?.content)) continue;
    for (const block of m.content as AnthropicBlock[]) {
      if (block?.type !== "tool_result" || !block.tool_use_id) continue;
      if (!markEmitted(`anthropic:${block.tool_use_id}`)) continue;
      emit(toolResultSpan("tool_result", block.content, !!block.is_error, block.tool_use_id));
    }
  }
}

// ─── Gemini (generateContent) ────────────────────────────────────────────────
interface GeminiPart { functionCall?: { name?: string; args?: unknown }; functionResponse?: { name?: string; response?: unknown; id?: string } }

/** Emit tool_call spans from Gemini response candidate parts (functionCall). */
export function emitGeminiToolCalls(candidates: unknown, parentId: string, model: string | undefined, emit: Emit): void {
  if (!Array.isArray(candidates)) return;
  for (const cand of candidates as Array<{ content?: { parts?: GeminiPart[] } }>) {
    const parts = cand?.content?.parts;
    if (!Array.isArray(parts)) continue;
    for (const p of parts) {
      if (!p.functionCall) continue;
      emit(toolCallSpan(p.functionCall.name || "tool", p.functionCall.args, parentId, model, p.functionCall.name));
    }
  }
}

/** Emit tool_result spans from Gemini request contents (functionResponse parts), deduped. */
export function emitGeminiToolResults(contents: unknown, emit: Emit): void {
  const list = Array.isArray(contents) ? contents : contents != null ? [contents] : [];
  for (const c of list as Array<{ parts?: GeminiPart[] }>) {
    const parts = c?.parts;
    if (!Array.isArray(parts)) continue;
    for (const p of parts) {
      const fr = p.functionResponse;
      if (!fr) continue;
      const key = `gemini:${fr.id || fr.name || JSON.stringify(fr.response).slice(0, 64)}`;
      if (!markEmitted(key)) continue;
      emit(toolResultSpan(fr.name || "tool_result", fr.response, false, fr.id || fr.name));
    }
  }
}

File: python/src/retrace/interceptors/tool_spans.py
"""Tool-span extraction (Phase 1C) for the Retrace Python SDK.

Provider interceptors historically emitted only a single ``llm_call`` span and dropped the
model's tool calls (the most common agent-failure class). These helpers derive structured
``tool_call`` and ``tool_result`` spans from a provider request/response so tool usage is
captured with no manual instrumentation:

- ``tool_call`` spans come from the model response (the calls it requested), arguments parsed
  into structured JSON.
- ``tool_result`` spans come from the tool outputs fed back on the next request (verbatim,
  including errors/empty), deduped by the provider tool-call id.

Downstream detectors (schema validation, loop detection, tool-output hallucination) rely on
these spans + the ``tool_call_id`` carried in metadata.
"""
import json
import uuid
from datetime import datetime, timezone

# Bounded dedup of emitted tool_result spans (keyed by provider tool-call id). Reset when a new
# trace installs its callback so memory stays bounded.
_emitted_tool_result_ids: set[str] = set()


def reset_tool_result_dedup() -> None:
    _emitted_tool_result_ids.clear()


def _mark_emitted(key: str) -> bool:
    if key in _emitted_tool_result_ids:
        return False
    if len(_emitted_tool_result_ids) > 5000:
        _emitted_tool_result_ids.clear()
    _emitted_tool_result_ids.add(key)
    return True


def _get(obj, key, default=None):
    """Read an attribute from either a dict or an SDK object."""
    if obj is None:
        return default
    if isinstance(obj, dict):
        return obj.get(key, default)
    return getattr(obj, key, default)


def _parse_args(args):
    if isinstance(args, str):
        try:
            return json.loads(args)
        except (ValueError, TypeError):
            return args
    return args


def extract_tool_schemas(provider: str, tools):
    """Extract declared tool parameter schemas (name → JSON schema) from a request's tools."""
    out: dict = {}
    if provider == "gemini":
        if not isinstance(tools, (list, tuple)):
            return None
        for group in tools:
            decls = _get(group, "function_declarations") or _get(group, "functionDeclarations") or []
            for fd in decls:
                name, params = _get(fd, "name"), _get(fd, "parameters")
                if name and params:
                    out[name] = params
    elif isinstance(tools, (list, tuple)):
        for t in tools:
            if provider == "openai":
                fn = _get(t, "function")
                name, params = _get(fn, "name"), _get(fn, "parameters")
                if name and params:
                    out[name] = params
            elif provider == "anthropic":
                name, schema = _get(t, "name"), _get(t, "input_schema")
                if name and schema:
                    out[name] = schema
    return out or None


def extract_sampling_params(provider: str, kwargs):
    """Capture the sampling/determinism envelope so replay-divergence (2A) + regression (2E) can
    compare sampling config, not just the model. Normalizes to
    {temperature, top_p, top_k, seed, max_tokens}. Returns None if none set."""
    cfg = _get(kwargs, "config") or {} if provider == "gemini" else (kwargs or {})
    out: dict = {}

    def put(k, v):
        if v is not None:
            out[k] = v

    if provider == "gemini":
        put("temperature", _get(cfg, "temperature"))
        put("top_p", _get(cfg, "top_p") or _get(cfg, "topP"))
        put("top_k", _get(cfg, "top_k") or _get(cfg, "topK"))
        put("seed", _get(cfg, "seed"))
        put("max_tokens", _get(cfg, "max_output_tokens") or _get(cfg, "maxOutputTokens"))
    else:
        put("temperature", _get(cfg, "temperature"))
        put("top_p", _get(cfg, "top_p"))
        put("top_k", _get(cfg, "top_k"))
        put("seed", _get(cfg, "seed"))
        put("max_tokens", _get(cfg, "max_tokens") or _get(cfg, "max_completion_tokens"))
    return out or None


def _truncate(val, limit: int = 4000):
    """Cap large string payloads; structured values are stored as-is (server uses jsonb)."""
    if isinstance(val, str):
        return val[:limit]
    return val


def _now_iso() -> str:
    return datetime.now(timezone.utc).isoformat().replace("+00:00", "Z")


def _tool_call_span(name, args, parent_id, model, tool_call_id=None) -> dict:
    now = _now_iso()
    d = {
        "id": str(uuid.uuid4()),
        "span_type": "tool_call",
        "name": name or "tool",
        "parent_id": parent_id,
        "input": _truncate(args),
        "started_at": now,
        "ended_at": now,
        "duration_ms": 0,
    }
    if model:
        d["model"] = model
    if tool_call_id:
        d["metadata"] = {"tool_call_id": tool_call_id}
    return d


def _tool_result_span(name, output, is_error, tool_call_id=None) -> dict:
    now = _now_iso()
    d = {
        "id": str(uuid.uuid4()),
        "span_type": "tool_result",
        "name": name or "tool_result",
        "output": _truncate(output),
        "started_at": now,
        "ended_at": now,
        "duration_ms": 0,
    }
    if is_error:
        d["error"] = output if isinstance(output, str) else json.dumps(output, default=str)
    if tool_call_id:
        d["metadata"] = {"tool_call_id": tool_call_id}
    return d


# ─── OpenAI ───────────────────────────────────────────────────────────────────
def emit_openai_tool_calls(message, parent_id, model, emit) -> None:
    calls = _get(message, "tool_calls")
    if not calls:
        return
    for c in calls:
        fn = _get(c, "function")
        emit(_tool_call_span(
            _get(fn, "name") or "tool", _parse_args(_get(fn, "arguments")),
            parent_id, model, _get(c, "id"),
        ))


def emit_openai_tool_results(messages, emit) -> None:
    if not isinstance(messages, (list, tuple)):
        return
    name_by_id: dict[str, str] = {}
    for m in messages:
        if _get(m, "role") == "assistant":
            for c in (_get(m, "tool_calls") or []):
                cid = _get(c, "id")
                if cid:
                    name_by_id[cid] = _get(_get(c, "function"), "name") or "tool"
    for m in messages:
        if _get(m, "role") != "tool":
            continue
        cid = _get(m, "tool_call_id")
        if not cid or not _mark_emitted(f"oa:{cid}"):
            continue
        content = _get(m, "content")
        is_error = isinstance(content, str) and any(w in content.lower() for w in ("error", "exception", "failed"))
        emit(_tool_result_span(name_by_id.get(cid) or _get(m, "name") or "tool_result", content, is_error, cid))


# ─── Anthropic ─────────────────────────────────────────────────────────────────
def emit_anthropic_tool_calls(content, parent_id, model, emit) -> None:
    if not isinstance(content, (list, tuple)):
        return
    for block in content:
        if _get(block, "type") != "tool_use":
            continue
        emit(_tool_call_span(_get(block, "name") or "tool", _get(block, "input"), parent_id, model, _get(block, "id")))


def emit_anthropic_tool_results(messages, emit) -> None:
    if not isinstance(messages, (list, tuple)):
        return
    for m in messages:
        content = _get(m, "content")
        if not isinstance(content, (list, tuple)):
            continue
        for block in content:
            if _get(block, "type") != "tool_result":
                continue
            tid = _get(block, "tool_use_id")
            if not tid or not _mark_emitted(f"anthropic:{tid}"):
                continue
            emit(_tool_result_span("tool_result", _get(block, "content"), bool(_get(block, "is_error")), tid))


# ─── Gemini ────────────────────────────────────────────────────────────────────
def emit_gemini_tool_calls(candidates, parent_id, model, emit) -> None:
    if not isinstance(candidates, (list, tuple)):
        return
    for cand in candidates:
        parts = _get(_get(cand, "content"), "parts") or []
        for p in parts:
            fc = _get(p, "function_call")
            if not fc:
                continue
            emit(_tool_call_span(_get(fc, "name") or "tool", _get(fc, "args"), parent_id, model, _get(fc, "name")))


def emit_gemini_tool_results(contents, emit) -> None:
    if contents is None:
        return
    items = contents if isinstance(contents, (list, tuple)) else [contents]
    for c in items:
        parts = _get(c, "parts")
        if not isinstance(parts, (list, tuple)):
            continue
        for p in parts:
            fr = _get(p, "function_response")
            if not fr:
                continue
            name = _get(fr, "name")
            resp = _get(fr, "response")
            key = f"gemini:{_get(fr, 'id') or name or str(resp)[:64]}"
            if not _mark_emitted(key):
                continue
            emit(_tool_result_span(name or "tool_result", resp, False, _get(fr, "id") or name))
